scala-native/scala-native
# Fix compiler plugin failures when compiling macros

This PR does fix issue reported in Scala Native discord channel 
```
there's a sporadic issue (I'm struggling to reproduce it 100%) when defining macros in a Scala 3 project, it seems to break compilation with SN:
----
sbt:tmp-jhdqlbbl> run
[info] compiling 2 Scala sources to /private/var/folders/my/drt5584s2w59ncgxgxpq7v040000gn/T/tmp.jHdqlBbL/target/scala-3.1.0/classes ...
java.lang.AssertionError: assertion failed: phase scalanative-prepareInterop has already been used once; cannot be reused while compiling /private/var/folders/my/drt5584s2w59ncgxgxpq7v040000gn/T/tmp.jHdqlBbL/Main.scala, /private/var/folders/my/drt5584s2w59ncgxgxpq7v040000gn/T/tmp.jHdqlBbL/bla.scala
[error] ## Exception when compiling 2 sources to /private/var/folders/my/drt5584s2w59ncgxgxpq7v040000gn/T/tmp.jHdqlBbL/target/scala-3.1.0/classes
[error] java.lang.AssertionError: assertion failed: phase scalanative-prepareInterop has already been used once; cannot be reused
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error] dotty.tools.dotc.core.Phases$Phase.init(Phases.scala:380)
[error] dotty.tools.dotc.core.Phases$Phase.init(Phases.scala:394)
---
Doesn't have to be even a good macro, this is enough to trigger this failure in a clean project:
---
object Macros:
  import scala.quoted.*

  inline def test(a: String): Seq[Int] = ${
    testCode('a)
  }
  private def testCode(s: Expr[String])(using Quotes): Expr[Seq[Int]] =
    import quotes.reflect.*
    Expr(List(1,2,3))
---
Problem is that it appears/disappears sometimes
```

* Changed definitions of all compiler plugin phases to be defined as `class` instead of `object`
* Make sure that that `NirDefinitions` are always defined using the current instance of context, by introducing `ContextCached` helper container
*  Add issue test reproducing the bug
